### DISCUSSION TITLE
谈谈 Kotlin 协程的 Context 和 Scope

### DISCUSSION THREAD

**Discussion by yujinyan** [yujinyan/blog#176] [Comments]
https://blog.yujinyan.me/posts/kotlin-coroutine-context-scope/

**Comment by Reign9201** [comment]
写的很不错，冒昧问下，你这画图工具是直接用iPad画图工具画的么

**Comment by yujinyan** [comment]
> 写的很不错，冒昧问下，你这画图工具是直接用iPad画图工具画的么

感谢留言😄~

插图用的是 https://excalidraw.com/ 这个工具。

**Comment by zombiu** [comment]
suspend 函数提供的约定：调用这个函数不会阻塞线程，函数内的子程序执行完毕以后函数才会返回，控制流回到调用方。suspend 函数不应该有开启异步任务的副作用。
这里说，suspend不会阻塞线程，又说函数执行完毕才会返回继续执行，这不是矛盾的么？

**Comment by yujinyan** [comment]
> suspend 函数提供的约定：调用这个函数不会阻塞线程，函数内的子程序执行完毕以后函数才会返回，控制流回到调用方。suspend 函数不应该有开启异步任务的副作用。
> 这里说，suspend不会阻塞线程，又说函数执行完毕才会返回继续执行，这不是矛盾的么？

@zombiu 

suspend 函数内部封装了回调，比如下面的例子：

```kotlin
suspend fun requestToken(): String
suspend fun createPost(token: String, item: Item): Post
suspend fun processPost(post)

suspend fun postItem(item: Item) {
  val token = 🏹 requestToken()
  val post = 🏹 createPost(token, item)
  🏹 processPost(post)
}
```

可以看成等价于这样回调的写法：

```kotlin
// 三个基于回调的 API
fun requestToken(block: (String) -> Unit)
fun createPost(token: String, item: Item, block: (Post) -> Unit))
fun processPost(post: Post)

fun postItem(item: Item) {
  requestToken { token ->
    createPost(token, item) { post ->
      processPost(post)
    }
  }
}
```

在这两个例子中三个步骤 `requestToken`，`createPost`，`processPost` 都是按个依次执行的，没有并发。所以说“函数内的子程序执行完毕以后函数才会返回，控制流回到调用方”。

假设 suspend 封装的是接口请求，比如 Retrofit (OkHttp)，最终会调用异步的 `Call.enqueue` 接口，在线程池里进行 http 请求，请求完成以后，控制流通过回调返回调用方。这样我们就可以直接在主线程调用 suspend 函数，函数执行完以后继续走后续的逻辑，但 suspend 函数执行的时候不会阻塞主线程，因为实际的请求在后台的线程池中进行，请求完成后通过回调回到主线程。

所以说要真正达到不阻塞调用方线程的效果，最终需要依赖异步的接口调度到别的线程去执行。

[《理解 Kotlin 的 suspend 函数》](https://blog.yujinyan.me/posts/understanding-kotlin-suspend-functions/) 这篇文章中有关于 suspend 更多介绍~

**Comment by liaolaolin** [comment]
文章写得很棒~

**Comment by ArnoFrost** [comment]
很详细，点赞~

**Comment by chr56** [comment]
写的真是很好，相当实用！

**Comment by etet2007** [comment]
写得可太好了 大佬
